### ISSUE TITLE
Raw printing not working with certain drivers

### ISSUE THREAD

**Issue by Arqit** [qzind/tray#763]
Hello, [](url)
When trying to raw print on a Lexmark b2338dw printer, the print jobs get sent to the que and the status remains as "sent to printer" indefinitely. When printing to a Lexmark MS810 Series XL printer, raw printing works. I decided to switch the driver of the former printer (after reading [this](https://h30434.www3.hp.com/t5/LaserJet-Printing/Send-Raw-Print-Jobs-do-not-print/m-p/6594229#M339470)) to match that of the latter and I managed to raw print successfully once the driver was changed. Is it possible to get driver independence?

**Comment by tresf** [comment]
> Is it possible to get driver independence?

You can use the "raw" driver, that's the most "independent" we have on Windows.  On CUPS, we offer something called "altPrinting", which bypasses the driver, but we haven't found a way to do this software side on Windows yet, so if the driver intercepts the raw commands, there's currently little we can do.

If you can find a way to bypass the printer driver on Windows, we'd be happy to add it as a new feature.

**Comment by Arqit** [comment]
Thank you, I'll give it a shot.  

To help, would you happen to have any intuition as to why raw printing to a printer with an incompatible driver using "echo this > Lexmark b2338dw" works but converting "this" to base64 and raw printing it through qz would just put the print job in the que indefinitely ?

**Comment by lite1979** [comment]
Arqit:

Can you tell us more about the raw language that you're sending to the Lexmark printer?  Their spec sheet states the following:


> Printer Languages | Standard:                     PCL 5e Emulation, PCL 6 Emulation, Microsoft XPS  (XML Paper Specification), Personal Printer Data Stream (PPDS),  PostScript 3 Emulation, Direct Image
> -- | --

The above info was gathered from https://www.lexmark.com/en_us/printer/11831/Lexmark-B2338dw on the "Specs" tab

"Emulation" generally means that the driver interprets any raw commands, not the printer itself, so creating a raw/generic queue that points to the printer port may not be a viable solution in this case. The drivers that offer raw language emulation are often Windows-only also.

With regards to using echo in command line working, and QZ not working. There are two possible points of failure due to encoding.

First, encoding to base64 may corrupt some characters before sending them

Second, sending via QZ-Tray without specifying the proper code page for Java can also influence the data in a way that prevents the stream from being interpreted correctly.

**Comment by tresf** [comment]
>  would you happen to have any intuition as to why raw printing to a printer with an incompatible driver using "echo this > Lexmark b2338dw" works but converting "this" to base64 and raw printing it through qz would just put the print job in the que indefinitely ?

* The first bypasses the driver completely.
* The second uses Java to print through the driver.

Requoting:

> You can use the "Generic Text/Raw" driver, that's the most "independent" we have on Windows. On CUPS, we offer something called `{ altPrinting: true }`, which bypasses the driver, but we haven't found a way to do this software side on Windows yet, so if the driver intercepts the raw commands, there's currently little we can do.

In addition, Lite's points about encoding are on-point, but those will largely depend on the content you're sending.  If you're just sending ASCII it's likely the driver stopping the raw data from going through, hence the need for a secondary printer which uses the Generic driver.  If you find a way to do this programmatically using Java or another technique we'd be happy to investigate.  In our experience 90% of Windows drivers allow the raw commands to go through, but we generally work with Epson and Zebra.  We have several clients using Lexmark, but we don't have one locally to test if the driver is interfering with the raw commands.

https://qz.io/wiki/setting-up-a-raw-printer-in-windows

**Comment by Arqit** [comment]
Thank you for the responses.
@lite1979 I attempted to print using EPL as well as ZPL. I see they don't appear on the list, I should have thought to check.  

Thanks for the suggested points of error but I don't believe they were the causes in my case as I was able to print the same base64 encoded string using QZ-tray with another printer.

@tresf If I manage to write code for a work around, I will make sure to share it. For now, changing to the "Generic Text/Raw" driver was able to solve the problem for me.
